jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,40 +191,115 @@ public int getMaxCapacity()
+        }
 
-                // A reserved entry may become old and be evicted before it is enabled.
-                if (reservedEntry.enable(buffer, true))
-                {
-                    if (direct)
-                        _currentDirectMemory.addAndGet(buffer.capacity());
-                    else
-                        _currentHeapMemory.addAndGet(buffer.capacity());
-                    releaseExcessMemory(direct);
-                    return buffer;
-                }
+        ByteBuffer byteBuffer = buffer.getByteBuffer();
+        BufferUtil.reset(byteBuffer);
+        Buffer pooledBuffer = new Buffer(byteBuffer, b -> release(bucket, entry));
+        // Discard the buffer if the entry cannot be enabled.
+        if (entry.enable(pooledBuffer, false))
+        {
+            updateMemory(bucket.getCapacity(), direct);
+        }
+        else
+        {
+            bucket._nonPooled.incrementAndGet();
+            entry.remove();
+        }
+    }
+
+    private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> entry)
+    {
+        bucket._releases.incrementAndGet();
+
+        RetainableByteBuffer buffer = entry.getPooled();
+        BufferUtil.reset(buffer.getByteBuffer());
+        long excessMemory = getExcessMemory(bucket, buffer.isDirect());
+        if (excessMemory > 0)
+        {
+            bucket._evicts.incrementAndGet();
+            // If we cannot free enough space for the entry, remove it.
+            if (!evict(excessMemory, buffer.isDirect()))
+            {
+                bucket._removes.incrementAndGet();
+                entry.remove();
+                return;
             }
-            return newRetainableByteBuffer(size, direct, this::removed);
         }
 
-        buffer = entry.getPooled();
-        ((Buffer)buffer).acquire();
-        return buffer;
+        // We have enough space for this entry, pool it.
+        if (!entry.release())
+        {
+            bucket._removes.incrementAndGet();
+            entry.remove();
+        }
+    }
+
+    private long getExcessMemory(RetainedBucket bucket, boolean direct)
+    {
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (maxMemory < 0)
+            return -1;
+        // Account also for the entry that is about to be released.
+        long memory = getMemory(direct) + bucket.getCapacity();
+        return memory - maxMemory;
+    }
+
+    private boolean evict(long excessMemory, boolean direct)
+    {
+        RetainedBucket[] buckets = direct ? _direct : _indirect;
+        int length = buckets.length;
+        int index = ThreadLocalRandom.current().nextInt(length);
+        for (int c = 0; c < length; ++c)
+        {
+            RetainedBucket bucket = buckets[index++];
+            if (index == length)
+                index = 0;
+            excessMemory -= bucket.evict();
+            if (excessMemory <= 0)
+                return true;
+        }
+        return false;
```

Comment:
This looks too complex.  Just pick a random bucket and if it is empty or the same size as the buffer we are releasing, then return false.    Sure it will miss a lot, but that is fine as we don't need to adapt quickly as size frequency changes are unlikely

```suggestion
    private boolean evict(long excessMemory, int releasedCapacity, boolean direct)
    {
        RetainedBucket[] buckets = direct ? _direct : _indirect;
        int length = buckets.length;
        int index = ThreadLocalRandom.current().nextInt(length);
        RetainedBucket bucket = buckets[index++];
        if (bucket == null || bucket.capacity() == releasedCapacity)
            return false;
        return bucket.evict() > excessMemory;
```

Reply:
@gregw I think your suggestion is just keeping idle buffers around, because now you have a random chance to hit the right bucket (1/N) to free a truly idle buffer.

Most of the time your suggestion will fail to free enough memory, returning `false` most of the time, and therefore discarding buffers that are actually in use (because they are being released back into the pool, but they won't).

Looping over all the buckets is not "too complex", it is 5 lines of code, and I think it's more fair.

If the "small" buckets are idle because the "large" buckets are in use, there is about 50% chance to start from smaller buckets and free one buffer each, and hopefully (if enough buffers are freed) keep the one being released.

If the "small" buckets are in use, there is still 50% chance to not overgrow them, and the idle "large" buffers get a 50% chance to be hit and slowly be removed, typically pooling the one being released (because it's smaller).

Reply:
Wow, it's like we've swapped brains and now you are arguing my points for a loop :)

The main thing I don't like about this version is the churning. It will often evict a buffer of size X only so it can add a buffer of size X.  At the very least it should never evict from the same bucket it is trying to add to.

I see the fact that my suggestion fails to evict 1/N times as a feature. Frequently, your original PR is correct and it is best to discard the just released buffer. So I'd say my suggestion is optimal (N-1)/N of the time as it will simply act like your original PR 

Changing size frequency is not the normal case, so a 1/N adaption rate should be sufficient to move to the new size(s) whilst not having lots of threads looping.

But I'm not going to die in a ditch about this. If you really want to loop, then loop. But please avoid the churn above and consider using an Atomic so only 1 thread is looping/evicting at the same time.

Reply:
The other aspect of this loop is that it is biased depending on the gaps between used buckets.   If we have the following buckets:
| Bucket  | Capacity | Size |  Probability | 
| ------------- | ------------- | -----| ------ |
| 0  | 2K  |  0 |  |
| 1  | 4K  |  20 | 2/16 |
| 2  | 6K  |  0 | |
| 3  | 8K  |  0 | |
| 4  | 10K  |  0 | |
| 5  | 12K  |  0 | |
| 6  | 14K  |  0 | |
| 7  | 16K  |  20 | 6/16 |
| 8  | 18K  |  20 | 1/16 |
| 9  | 20K  |  0 | |
| 10 | 22K  |  0 | |
| 11 | 24K  |  0 | |
| 12 | 26K  |  0 | |
| 13 | 28K  |  0 | |
| 14 | 30K  |  0 | |
| 15 | 32K  |  20 | 7/16 |

Then the 32K buckets are 7 times more likely to be evicted than the 18K buckets, and 3.5 times more likely than the 4K buckets.

I'd much rather just evict 3/16 of the time and keep it unbiased.

Reply:
@gregw you make good points, but I don't think we have enough data to decide.
Let's discuss this in person, and/or when we have more data.

Reply:
OK, but I will type a few more thoughts here so I don't forget before we can hangout.

From a small sample, we "typically" have 16 buckets of which 4 to 6 sizes are used, "typically" 2 or 3 of those have a lot more usage than the others.

If we are release a buffer that will exceed the max size then either:
 + it's bucket already has some idle buffers of the same size, so we can just discard this buffer.
 + it's bucket has no idle buffers, so we need to evict from another bucket picked at random:
    * 1/16 chance that it picks the same bucket, so give up and discard.  We might be unbalanced but chances are a future release will fix it.
    * 10/16 chance that it picks an empty bucket, so give up and discard. We might be unbalanced but chances are a future release will fix it.
    * 5/16 chance that a non empty bucket is picked, which might be:
      - not enough space is evicted for the released buffer, so clear what we can and discard.  We might be unbalanced, but have freed some space and future releases can improve.
      - enough space is evicted, so the released buffer is added and we are probably more balanced
 
So more often than not, this will just discard the buffer when max capacity is exceeded, which is simple, fast and probably the best thing to do in almost all circumstances... unless we are unbalanced due to a size frequency change... in which case there is a reasonable chance (2.5/16) that we will move towards being more balanced.  The release rate is such that we will quickly rebalance even with a small chance.